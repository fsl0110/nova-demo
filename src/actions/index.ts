import openFDA from "../api/openFDA";
import { Dispatch } from "redux";
import {
  FETCH_REPORTS_START,
  FETCH_REPORTS_END,
  FETCH_REPORTS_ERROR,
  FETCH_REPORTS_RESET,
  FETCH_INITIATOR_START,
  FETCH_INITIATOR_END,
  FETCH_INITIATOR_ERROR
} from "../typings";

// Reports Actions
export const fetchReports = (term: string = "") => async (
  dispatch: Dispatch
) => {
  dispatch(fetchReportsStart());
  openFDA
    .get(
      `food/enforcement.json?search=reason_for_recall:"${term}"&count=report_date`
    )
    .then(res => {
      dispatch(
        dispatch({ type: FETCH_REPORTS_END, payload: res.data.results })
      );
    })
    .catch(err => {
      if (!err.response || err.response.status !== 404) {
        dispatch(fetchReportsError());
      } else {
        dispatch(fetchReportsReset());
      }
    });
};

export const fetchReportsStart = () => {
  return {
    type: FETCH_REPORTS_START
  };
};

export const fetchReportsError = () => {
  return {
    type: FETCH_REPORTS_ERROR
  };
};

export const fetchReportsReset = () => {
  return {
    type: FETCH_REPORTS_RESET
  };
};

// Initiator Actions

export const fetchInitiator = () => async (dispatch: Dispatch) => {
  dispatch(fetchInitiatorStart());
  openFDA
    .get(`food/enforcement.json?count=voluntary_mandated.exact`)
    .then(res => {
      dispatch(
        dispatch({ type: FETCH_INITIATOR_END, payload: res.data.results })
      );
    })
    .catch(err => {
      dispatch(fetchInitiatorError());
    });
};

export const fetchInitiatorStart = () => {
  return {
    type: FETCH_INITIATOR_START
  };
};

export const fetchInitiatorError = () => {
  return {
    type: FETCH_INITIATOR_ERROR
  };
};
