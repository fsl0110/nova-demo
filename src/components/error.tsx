import React, { FC, HTMLAttributes } from "react";
import { Button } from "../components/button";

export interface Props extends HTMLAttributes<HTMLDivElement> {
  /** Is loading data */
  isLoading: boolean;
  /** ClickHandler for fetching data */
  fetchData: () => void;
}

export const Error: FC<Props> = ({ isLoading, fetchData }) => (
  <div className="error">
    <div className="text-center">
      <h4>An error occured</h4>
      <p>This Chart can't be shown. Click the reload button to try again.</p>
      <Button
        aria-label="Reloadbutton for reloading Chartdatas"
        className="error__button"
        iconName="refresh"
        iconAlign="after"
        iconSize="m"
        isLoading={isLoading}
        isLoadingSpin={true}
        onClick={() => fetchData()}
      >
        {isLoading ? "Loading" : "Reload"}
      </Button>
    </div>
  </div>
);

export default Error;
