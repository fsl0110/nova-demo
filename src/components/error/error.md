### Display fetching errors

#### With isLoading=false

```jsx
<div className="food">
  <div className="chart-wrapper">
    <Error isLoading={false} fetchData={() => null} />
  </div>
</div>
```

#### With isLoading=true

```jsx
<div className="chart-wrapper">
  <Error isLoading={true} fetchData={() => null} />
</div>
```
