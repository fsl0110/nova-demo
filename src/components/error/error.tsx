import React, { FC, HTMLAttributes } from "react";
import { Button } from "..";

export interface Props extends HTMLAttributes<HTMLDivElement> {
  /** Is loading data */
  isLoading?: boolean;
  /** ClickHandler for fetching data */
  fetchData: () => void;
}

export const Error: FC<Props> = ({ isLoading, fetchData }) => (
  <div className="error" data-test="Error">
    <div className="text-center">
      <h4>An error occured</h4>
      <p>This Chart can't be shown. Click the reload button to try again.</p>
      <Button
        aria-label="Reloadbutton for reloading Chartdatas"
        className="error__button"
        iconName="refresh"
        iconAlign="after"
        iconSize="m"
        isLoading={isLoading}
        isLoadingSpin={true}
        onClick={() => fetchData()}
        data-test="error-button"
      >
        {isLoading ? "Loading" : "Reload"}
      </Button>
    </div>
  </div>
);

export default Error;
