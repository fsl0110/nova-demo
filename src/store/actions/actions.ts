import { axiosOpenFDA } from "../../utils/api/openFDA";
import { Dispatch } from "redux";
import moment from "moment";

export const FETCH_START = "FETCH_START";
export const FETCH_SUCCESS = "FETCH_SUCCESS";
export const FETCH_ERROR = "FETCH_ERROR";
export const FETCH_RESET = "FETCH_RESET";

export type ActionTypes = FetchStart | FetchSuccess | FetchError | FetchReset;

export interface FetchSuccess {
  type: typeof FETCH_SUCCESS;
  payload: any;
  location: string;
}

export const fetchReports = (term: string = "") => (dispatch: Dispatch) => {
  dispatch(fetchStart("reports"));
  axiosOpenFDA({
    method: "get",
    responseType: "json",
    url: `food/enforcement.json?search=reason_for_recall:"${term}"&count=report_date`,
    transformResponse: r =>
      r.results.map((i: { time: number; count: number }) => [
        moment(i.time, "YYYY-MM-DD").valueOf(),
        i.count
      ])
  })
    .then(res => {
      dispatch(
        dispatch({
          type: FETCH_SUCCESS,
          payload: res.data,
          location: "reports"
        })
      );
    })
    .catch(err => {
      if (!err.response || err.response.status !== 404) {
        dispatch(fetchError("reports"));
      } else {
        dispatch(fetchReset("reports"));
      }
    });
};

export const fetchInitiators = () => (dispatch: Dispatch) => {
  dispatch(fetchStart("initiators"));
  axiosOpenFDA({
    method: "get",
    responseType: "json",
    url: `food/enforcement.json?count=voluntary_mandated.exact`,
    transformResponse: r =>
      r.results.map((i: { term: string; count: number }) => [i.term, i.count])
  })
    .then(res => {
      dispatch(
        dispatch({
          type: FETCH_SUCCESS,
          payload: res.data,
          location: "initiators"
        })
      );
    })
    .catch(err => {
      dispatch(fetchError("initiators"));
    });
};

export interface FetchStart {
  type: typeof FETCH_START;
  location: string;
}

export const fetchStart = (location: string): FetchStart => {
  return {
    type: FETCH_START,
    location
  };
};

export interface FetchError {
  type: typeof FETCH_ERROR;
  location: string;
}

export const fetchError = (location: string): FetchError => {
  return {
    type: FETCH_ERROR,
    location
  };
};

export interface FetchReset {
  type: typeof FETCH_RESET;
  location: string;
}

export const fetchReset = (location: string) => {
  return {
    type: FETCH_RESET,
    location
  };
};
