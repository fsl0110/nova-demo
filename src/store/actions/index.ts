import openFDA from "../../setttings/api/openFDA";
import { Dispatch } from "redux";

export const FETCH_START = "FETCH_START";
export const FETCH_SUCCESS = "FETCH_SUCCESS";
export const FETCH_ERROR = "FETCH_ERROR";
export const FETCH_RESET = "FETCH_RESET";

export type ActionTypes = FetchStart | FetchSuccess | FetchError | FetchReset;

export interface FetchSuccess {
  type: typeof FETCH_SUCCESS;
  payload: [];
  location: string;
  term?: string;
}

export const fetchData = (location: string, term: string = ""): any => (
  dispatch: Dispatch
) => {
  dispatch(fetchStart(location));
  openFDA
    .get(
      location === "reports"
        ? `food/enforcement.json?search=reason_for_recall:"${term}"&count=report_date`
        : `food/enforcement.json?count=voluntary_mandated.exact`
    )
    .then(res => {
      dispatch(
        dispatch({ type: FETCH_SUCCESS, payload: res.data.results, location })
      );
    })
    .catch(err => {
      if (!err.response || err.response.status !== 404) {
        dispatch(fetchError(location));
      } else {
        dispatch(fetchReset(location));
      }
    });
};

export interface FetchStart {
  type: typeof FETCH_START;
  location: string;
}

export const fetchStart = (location: string): FetchStart => {
  return {
    type: FETCH_START,
    location
  };
};

export interface FetchError {
  type: typeof FETCH_ERROR;
  location: string;
}

export const fetchError = (location: string): FetchError => {
  return {
    type: FETCH_ERROR,
    location
  };
};

export interface FetchReset {
  type: typeof FETCH_RESET;
  location: string;
}

export const fetchReset = (location: string): FetchReset => {
  return {
    type: FETCH_RESET,
    location
  };
};

/* 
export const fetchInitiators = () => (dispatch: Dispatch) => {
  dispatch(fetchInitiatorsStart());
  openFDA
    .get(`food/enforcement.json?count=voluntary_mandated.exact`)
    .then(res => {
      dispatch(
        dispatch({ type: FETCH_INITIATORS_SUCCESS, payload: res.data.results })
      );
    })
    .catch(err => {
      dispatch(fetchInitiatorsError());
    });
};

export const fetchInitiatorsStart = () => {
  return {
    type: FETCH_INITIATORS_START
  };
};

export const fetchInitiatorsError = () => {
  return {
    type: FETCH_INITIATORS_ERROR
  };
};
 */
