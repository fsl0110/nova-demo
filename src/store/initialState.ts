import { AppState } from "../typings";
import { hcGlobal } from "../static/charts/hcGlobal";
import { hcEnforcementReports } from "../static/charts/hcEnforcementReports";
import { hcEnforcementInitiators } from "../static/charts/hcEnforcementInitiators";

export const initialState: AppState = {
  reports: {
    chartOptions: { ...hcGlobal, ...hcEnforcementReports },
    isLoading: false,
    hasError: false
  },
  initiators: {
    chartOptions: { ...hcGlobal, ...hcEnforcementInitiators },
    isLoading: false,
    hasError: false
  }
};
