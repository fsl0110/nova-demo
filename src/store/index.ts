import { createStore, applyMiddleware } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import thunk from "redux-thunk";
import { reducer } from "./reducers/reducers";
import { initialState } from "./initialState";

export const middlewares = [thunk];

const store = createStore(
  reducer,
  initialState,
  composeWithDevTools(applyMiddleware(...middlewares))
);

export default store;
