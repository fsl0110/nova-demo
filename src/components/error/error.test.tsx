import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import configureStore from "redux-mock-store";
import { initialState } from "../../store/initialState";
import { Error, Props } from "./error";
import { fetchData, fetchStart } from "../../store/actions/actions";

const testCases: Array<[string, Props]> = [
  [
    "Error while loading",
    { location: "reports", isLoading: true, fetchData: jest.fn() }
  ],
  [
    "Error default",
    { location: "reports", isLoading: false, fetchData: jest.fn() }
  ]
];

describe("<Error>", () => {
  testCases.forEach(([item, options]) => {
    it(item, () => {
      expect(toJson(shallow(<Error {...options} />))).toMatchSnapshot();
    });
  });

  it("should handle the click event", () => {
    /*     const mockStore = configureStore();
    const store = mockStore(initialState);
    store.dispatch(fetchStart("reports"));

    const mock = shallow(
      <Error location="reports" fetchData={fetchStart("reports")}>
        Some text
      </Error>
    );
    mock.simulate("click");
    expect(window.alert).toHaveBeenCalledWith("clicked"); */
  });
});
