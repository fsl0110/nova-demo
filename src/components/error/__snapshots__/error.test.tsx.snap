// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Error> Error default 1`] = `
<div
  className="error"
  data-test="Error"
>
  <div
    className="text-center"
  >
    <h4>
      An error occured
    </h4>
    <p>
      This Chart can't be shown. Click the reload button to try again.
    </p>
    <Button
      aria-label="Reloadbutton for reloading Chartdatas"
      className="error__button"
      iconAlign="after"
      iconName="refresh"
      iconSize="m"
      isLoading={false}
      isLoadingSpin={true}
      onClick={[Function]}
    >
      Reload
    </Button>
  </div>
</div>
`;

exports[`<Error> Error while loading 1`] = `
<div
  className="error"
  data-test="Error"
>
  <div
    className="text-center"
  >
    <h4>
      An error occured
    </h4>
    <p>
      This Chart can't be shown. Click the reload button to try again.
    </p>
    <Button
      aria-label="Reloadbutton for reloading Chartdatas"
      className="error__button"
      iconAlign="after"
      iconName="refresh"
      iconSize="m"
      isLoading={true}
      isLoadingSpin={true}
      onClick={[Function]}
    >
      Loading
    </Button>
  </div>
</div>
`;
